src/misc/rwqueue.cpp

```diff
@@ -105,42 +142,58 @@ void RWQueue<T>::BulkEnqueue(std::vector<T>& from_source, const size_t num_reque
 		                                std::min(num_remaining,
 		                                         free_capacity));
 
-		// wait until the queue has enough room for the items
-		has_room.wait(lock, [&] { return capacity - queue.size() >= num_items; });
-
-		const auto source_end = source_start +
-		                        static_cast<difference_t>(num_items);
-		queue.insert(queue.end(),
-		             std::move_iterator(source_start),
-		             std::move_iterator(source_end));
+		// wait until we're stopped or the queue has enough room for the
+		// items
+		has_room.wait(lock, [&] {
+			return is_stopped || capacity - queue.size() >= num_items;
+		});
+
+		if (!is_stopped) {
+			const auto source_end = source_start +
+			                        static_cast<difference_t>(num_items);
+			queue.insert(queue.end(),
+			             std::move_iterator(source_start),
+			             std::move_iterator(source_end));
+			source_start = source_end;
+			num_remaining -= num_items;
+		} else {
+			// If we stopped while bulk enqueing, then stop here.
+			// Anything that was enqueued prior to being stopped is
+			// safely in the queue.
+			num_remaining = 0;
```

Comment:
I think that's a sensible choice. It should be the client's responsibility to defer the queue being stopped if it wants to ensure some (bulk)enqueue operations are fully processed. This makes for a more flexible API too.